# Python Midterm Exam

### Sander Belon

### 14.11.2022

## Problem 1

In [4]:
#importing packages

import pandas as pd
import numpy as np
import requests
import json

In [5]:
url = 'https://ies-python-midterm.s3.eu-central-1.amazonaws.com/studentsSets/30241262.zip'

In [6]:
r = requests.get(url)

In [60]:
L = ('./data/L.csv')

In [61]:
C = ('./data/C.csv')

In [62]:
ADP = ('./data/ADP.csv')

In [63]:
LUV = ('./data/LUV.csv')

In [49]:
df = pd.concat(
    map(pd.read_csv, [df_L, df_ADP, df_C, df_LUV]))

In [70]:
df_L = pd.DataFrame(pd.read_csv(L))

In [71]:
df_L = df_L.set_index('Date')

In [80]:
df_C = pd.DataFrame(pd.read_csv(C))

In [81]:
df_C = df_C.set_index('Date')

In [83]:
df_ADP = pd.DataFrame(pd.read_csv(ADP))

In [84]:
df_ADP = df_ADP.set_index('Date')

In [87]:
df_LUV = pd.DataFrame(pd.read_csv(LUV))

In [88]:
df_LUV = df_LUV.set_index('Date')

In [91]:
df_L['Close']!=df_L['Adj Close']

Date
2017-11-13     True
2017-11-14     True
2017-11-15     True
2017-11-16     True
2017-11-17     True
              ...  
2022-11-07    False
2022-11-08    False
2022-11-09    False
2022-11-10    False
2022-11-11    False
Length: 1259, dtype: bool

In [93]:
df_C['Close']!=df_C['Adj Close']

Date
2017-11-13     True
2017-11-14     True
2017-11-15     True
2017-11-16     True
2017-11-17     True
              ...  
2022-11-07    False
2022-11-08    False
2022-11-09    False
2022-11-10    False
2022-11-11    False
Length: 1259, dtype: bool

In [95]:
df_LUV['Close']!=df_LUV['Adj Close']

Date
2017-11-13     True
2017-11-14     True
2017-11-15     True
2017-11-16     True
2017-11-17     True
              ...  
2022-11-07    False
2022-11-08    False
2022-11-09    False
2022-11-10    False
2022-11-11    False
Length: 1259, dtype: bool

In [96]:
df_ADP['Close']!=df_ADP['Adj Close']

Date
2017-11-13     True
2017-11-14     True
2017-11-15     True
2017-11-16     True
2017-11-17     True
              ...  
2022-11-07    False
2022-11-08    False
2022-11-09    False
2022-11-10    False
2022-11-11    False
Length: 1259, dtype: bool

Yes, all companies experience at a specific date a difference between Close and Adj Close. This means that ...

In [98]:
df_C.describe()

High          Low         Open        Close        Volume  \
count  1259.000000  1259.000000  1259.000000  1259.000000  1.259000e+03   
mean     63.872065    62.380485    63.152280    63.106052  1.971048e+07   
std      10.615420    10.769412    10.685572    10.694397  8.784308e+06   
min      37.990002    32.000000    35.470001    35.389999  4.576200e+06   
25%      53.635000    52.025000    52.820002    52.809999  1.356420e+07   
50%      66.830002    65.230003    66.129997    66.080002  1.766090e+07   
75%      72.000000    70.680000    71.334999    71.375000  2.389980e+07   
max      83.110001    80.970001    81.959999    81.910004  8.656620e+07   

         Adj Close  
count  1259.000000  
mean     59.258520  
std       9.388381  
min      32.968502  
25%      51.440001  
50%      60.896164  
75%      66.410423  
max      78.082741

In [99]:
df_L.describe()

High          Low         Open        Close        Volume  \
count  1259.000000  1259.000000  1259.000000  1259.000000  1.259000e+03   
mean     50.671660    49.703312    50.184075    50.176052  1.057290e+06   
std       7.579851     7.681085     7.633258     7.632317  5.437245e+05   
min      29.420000    27.330000    28.250000    28.799999  3.001000e+05   
25%      48.025000    47.205000    47.594999    47.575001  7.223000e+05   
50%      51.049999    50.299999    50.689999    50.759998  9.347000e+05   
75%      55.539999    54.625000    55.094999    55.119999  1.222200e+06   
max      68.199997    66.599998    67.500000    67.360001  5.592500e+06   

         Adj Close  
count  1259.000000  
mean     49.702547  
std       7.721808  
min      28.493643  
25%      46.878710  
50%      49.952072  
75%      54.881210  
max      67.360001

In [100]:
df_LUV.describe()

High          Low         Open        Close        Volume  \
count  1259.000000  1259.000000  1259.000000  1259.000000  1.259000e+03   
mean     49.669531    48.304702    49.010350    48.968205  7.298895e+06   
std       8.913654     9.044236     8.970799     8.984196  6.304535e+06   
min      24.809999    22.469999    23.379999    23.870001  1.098400e+06   
25%      43.490002    41.990000    42.755001    42.820000  4.100250e+06   
50%      51.560001    50.340000    50.959999    50.910000  5.683900e+06   
75%      55.889999    54.540001    55.260000    55.115000  8.255000e+06   
max      66.989998    65.980003    66.500000    66.290001  1.010078e+08   

         Adj Close  
count  1259.000000  
mean     48.537834  
std       8.656748  
min      23.870001  
25%      42.820000  
50%      50.234112  
75%      54.644875  
max      64.447029

In [101]:
df_ADP.describe()

High          Low         Open        Close        Volume  \
count  1259.000000  1259.000000  1259.000000  1259.000000  1.259000e+03   
mean    172.437522   169.084520   170.771398   170.861612  1.978432e+06   
std      37.752878    37.177623    37.455614    37.463405  1.039699e+06   
min     110.440002   103.110001   108.529999   108.250000  3.502000e+05   
25%     142.335007   139.080002   140.985001   140.805000  1.364000e+06   
50%     165.619995   162.710007   164.410004   164.389999  1.739700e+06   
75%     204.384995   200.279999   202.160004   201.989998  2.264100e+06   
max     261.589996   259.549988   260.369995   260.940002  1.711080e+07   

         Adj Close  
count  1259.000000  
mean    163.661444  
std      40.471893  
min      98.127808  
25%     132.173737  
50%     154.279160  
75%     198.518616  
max     259.817322

- For company C:
-- The highest price for Adj Close counts 78.082741.
-- The lowes price counts 32.968502.
- For company L:
-- The highest price for Adj Close counts 67.360001.
-- The lowes price counts 28.493643.
- For company LUV:
-- The highest price for Adj Close counts 64.447029.
-- The lowes price counts 23.870001.
- For company ADP:
-- The highest price for Adj Close counts 259.817322.
-- The lowes price counts 98.127808.

In [104]:
df_L['ln']= np.log2(df_L['Adj Close'])
df_C['ln']= np.log2(df_C['Adj Close'])
df_LUV['ln']= np.log2(df_LUV['Adj Close'])
df_ADP['ln']= np.log2(df_ADP['Adj Close'])

In [108]:
df_L['ln'].describe()

count    1259.000000
mean        5.616136
std         0.241265
min         4.832568
25%         5.550861
50%         5.642473
75%         5.778240
max         6.073820
Name: ln, dtype: float64

In [112]:
df_C['ln'].describe()

count    1259.000000
mean        5.869359
std         0.243051
min         5.043016
25%         5.684819
50%         5.928279
75%         6.053338
max         6.286932
Name: ln, dtype: float64

In [113]:
df_LUV['ln'].describe()

count    1259.000000
mean        5.575712
std         0.278040
min         4.577127
25%         5.420213
50%         5.650595
75%         5.772014
max         6.010042
Name: ln, dtype: float64

In [114]:
df_ADP['ln'].describe()

count    1259.000000
mean        7.311148
std         0.353637
min         6.616590
25%         7.046292
50%         7.269399
75%         7.633130
max         8.021354
Name: ln, dtype: float64

In [119]:
df_L.idxmax()

High         2022-04-21
Low          2022-04-21
Open         2022-04-21
Close        2022-04-20
Volume       2021-09-17
Adj Close    2022-04-20
Match        2017-11-13
ln           2022-04-20
dtype: object

In [123]:
df_C.idxmax()

High         2020-01-14
Low          2020-01-14
Open         2020-01-16
Close        2020-01-14
Volume       2022-07-15
Adj Close    2021-06-02
ln           2021-06-02
dtype: object

In [124]:
df_LUV.idxmax()

High         2017-12-21
Low          2017-12-21
Open         2017-12-21
Close        2018-01-02
Volume       2020-04-29
Adj Close    2018-01-02
ln           2018-01-02
dtype: object

In [125]:
df_ADP.idxmax()

High         2022-08-18
Low          2022-08-18
Open         2022-08-18
Close        2022-08-18
Volume       2019-07-02
Adj Close    2022-08-18
ln           2022-08-18
dtype: object

## Problem 2

In [133]:
url = 'https://ies-python-midterm.s3.eu-central-1.amazonaws.com/companies/companies_no_subindustry.csv'
r = requests.get(url)



'Symbol,Security,GICSÂ\xa0Sector,hq,included,Founded\r\nMMM,3M,Industrials,"Saint Paul, Minnesota",09.08.1976,1902\r\nAOS,A. O. Smith,Industrials,"Milwaukee, Wisconsin",26.07.2017,1916\r\nABT,Abbott,Health Care,"North Chicago, Illinois",31.03.1964,1888\r\nABBV,AbbVie,Health Care,"North Chicago, Illinois",31.12.2012,2013 (1888)\r\nABMD,Abiomed,Health Care,"Danvers, Massachusetts",31.05.2018,1981\r\nACN,Accenture,Information Technology,"Dublin, Ireland",06.07.2011,1989\r\nATVI,Activision Blizzard,Communication Services,"Santa Monica, California",31.08.2015,2008\r\nADM,ADM,Consumer Staples,"Chicago, Illinois",29.07.1981,1902\r\nADBE,Adobe Inc.,Information Technology,"San Jose, California",05.05.1997,1982\r\nADP,ADP,Information Technology,"Roseland, New Jersey",31.03.1981,1949\r\nAAP,Advance Auto Parts,Consumer Discretionary,"Raleigh, North Carolina",09.07.2015,1932\r\nAES,AES Corporation,Utilities,"Arlington, Virginia",02.10.1998,1981\r\nAFL,Aflac,Financials,"Columbus, Georgia",28.05.1